### PR TITLE
Support vip

### PR DESCRIPTION
**Pull request by yaocw2020**
Add vip controller to support vip.

issue: https://github.com/harvester/harvester/issues/633

### CODE REVIEW

File: pkg/controller/common/nodenetwork.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ import (
 )
 
 const (
+	Namespace        = "harvester-system"
```
Comment (guangbochen) [inline pkg/controller/common/nodenetwork.go:17]: why don't we get ns from https://github.com/harvester/network-controller-harvester/blob/master/main.go#L47

File: pkg/controller/manager/setup.go [hunk 2]
```diff
@@ -4,9 +4,11 @@ import (
 	"github.com/harvester/harvester-network-controller/pkg/config"
 	"github.com/harvester/harvester-network-controller/pkg/controller/manager/clusternetwork"
 	"github.com/harvester/harvester-network-controller/pkg/controller/manager/node"
+	"github.com/harvester/harvester-network-controller/pkg/controller/manager/vip"
 )
 
 var RegisterFuncList = []config.RegisterFunc{
 	clusternetwork.Register,
 	node.Register,
+	vip.Register,
```
Comment (guangbochen) [inline pkg/controller/manager/setup.go:13]: prefer to add a common option to either enable or disable the VIP integration via https://github.com/harvester/network-controller-harvester/blob/master/main.go#L33

File: pkg/controller/manager/vip/controller.go [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+	var vipInterface string
+	var containerIndex, envIndex int
+	for i, c := range ds.Spec.Template.Spec.Containers {
+		if c.Name == vipContainerName {
+			containerIndex = i
+			for j, env := range c.Env {
+				if env.Name == vipInterfaceEnvName {
+					envIndex = j
+					vipInterface = env.Value
```
Comment (guangbochen) [inline pkg/controller/manager/vip/controller.go:115]: should break once found it

File: pkg/controller/manager/vip/controller.go [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	if nn.Spec.Type != networkv1.NetworkTypeVLAN {
+		return nn, nil
+	}
+
+	for _, con := range nn.Status.Conditions {
+		if con.Type == networkv1.NodeNetworkRemoved {
+			if con.Status == corev1.ConditionTrue {
+				return nn, h.updateVipInterface(vlan.BridgeName)
+			}
+			return nil, fmt.Errorf("the vlan network hasn't removed")
```
Comment (guangbochen) [inline pkg/controller/manager/vip/controller.go:91]: should return this at the end of func.

File: pkg/controller/common/nodenetwork.go (hunk 1 above)
Comment (yaocw2020) [reply pkg/controller/common/nodenetwork.go:17]: It's the namespace where kube-vip deploys.

Comment (guangbochen) [reply pkg/controller/common/nodenetwork.go:17]: /s/Namespace/KubeVipNamespace/

File: manifests/deployment.yaml [hunk 5]
```diff
@@ -58,11 +58,12 @@ spec:
       containers:
         - image: rancher/harvester-network-controller:master-head
           imagePullPolicy: IfNotPresent
-          name: network
+          name: harvester-manager-network-controller
```
Comment (guangbochen) [inline manifests/deployment.yaml:61]: Nit: shorter is better since this is the container name

File: manifests/deployment.yaml [hunk 6]
```diff
@@ -58,11 +58,12 @@ spec:
...
           command:
             - harvester-network-controller
           args:
             - manager
+            - --enable-vip-controller
```
Comment (guangbochen) [inline manifests/deployment.yaml:66]: default to false, therefore this is useless, you can enable it on the harvester side.

File: pkg/controller/manager/setup.go [hunk 7]
```diff
@@ -1,12 +1,25 @@
...
 	"github.com/harvester/harvester-network-controller/pkg/config"
 	"github.com/harvester/harvester-network-controller/pkg/controller/manager/clusternetwork"
 	"github.com/harvester/harvester-network-controller/pkg/controller/manager/node"
+	"github.com/harvester/harvester-network-controller/pkg/controller/manager/vip"
 )
 
-var RegisterFuncList = []config.RegisterFunc{
-	clusternetwork.Register,
-	node.Register,
+var registerMap = map[string]config.RegisterFunc{
```
Comment (gitlawr) [inline pkg/controller/manager/setup.go:12]: I think we don't need to change RegisterFuncList to a map. We can append `vip.Register` to the funcList when `--enable-vip-controller` is set

File: pkg/controller/manager/vip/controller.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+		}
+	}
+
+	if vipInterface != unavailableNIC {
+		return nil
+	}
+
+	dsCopy := ds.DeepCopy()
+	if vipInterface == vlan.BridgeName {
+		dsCopy.Spec.Template.Spec.Containers[containerIndex].Env[envIndex].Value = defaultVipInterface
```
Comment (gitlawr) [inline pkg/controller/manager/vip/controller.go:129]: We are using eth0 as a fixed default. Should it be configurable?

File: pkg/controller/manager/vip/controller.go [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+	} else {
+		dsCopy.Spec.Template.Spec.Containers[containerIndex].Env[envIndex].Value = vlan.BridgeName
+	}
+
+	if _, err := h.dsCtr.Update(dsCopy); err != nil {
```
Comment (gitlawr) [inline pkg/controller/manager/vip/controller.go:134]: Will this env change get overridden on helm upgrade?

Comment (yaocw2020) [reply pkg/controller/manager/vip/controller.go:143]: Only in the case of upgrading kube-vip

**Review by gitlawr — APPROVED** [review]
LGTM

**Review by guangbochen — APPROVED** [review]
LGTM

**Pull request merged**